torchrl/objectives/costs/dreamer.py

```diff
@@ -0,0 +1,289 @@
+                [],
+            ),
+            tensordict.detach(),
+        )
+
+    def kl_loss(
+        self,
+        prior_mean: torch.Tensor,
+        prior_std: torch.Tensor,
+        posterior_mean: torch.Tensor,
+        posterior_std: torch.Tensor,
+    ) -> torch.Tensor:
+        kl = (
+            torch.log(prior_std / posterior_std)
+            + (posterior_std ** 2 + (prior_mean - posterior_mean) ** 2)
+            / (2 * prior_std ** 2)
+            - 0.5
+        )
+        if self.inversed_free_nats:
+            kl = kl.sum(-1).mean().clamp_min(self.free_nats)
+        else:
+            kl = kl.sum(-1).clamp_min(self.free_nats).mean()
+        return kl
+
+
+class DreamerActorLoss(LossModule):
+    """Dreamer Actor Loss
+
+    Computes the loss of the dreamer actor. The actor loss is computed as the negative average lambda return.
+
+    Reference: https://arxiv.org/abs/1912.01603
+
+    Args:
+        actor_model (TensorDictModule): the actor model.
+        value_model (TensorDictModule): the value model.
+        model_based_env (DreamerEnv): the model based environment.
+        imagination_horizon (int, optional): The number of steps to unroll the model. default: 15.
+        gamma (float, optional): the gamma discount factor. default: 0.99.
+        lmbda (float, optional): the lambda discount factor factor. default: 0.95
+        discount_loss (bool, optional): if True, the loss is discounted with a gamma discount factor. default: True.
+    """
+
+    def __init__(
+        self,
+        actor_model: TensorDictModule,
+        value_model: TensorDictModule,
+        model_based_env: DreamerEnv,
+        imagination_horizon: int = 15,
+        gamma: int = 0.99,
+        lmbda: int = 0.95,
+        discount_loss: bool = True,
+    ):
+        super().__init__()
+        self.actor_model = actor_model
+        self.value_model = value_model
+        self.model_based_env = model_based_env
+        self.imagination_horizon = imagination_horizon
+        self.gamma = gamma
+        self.lmbda = lmbda
+        self.discount_loss = discount_loss
+
+    def forward(self, tensordict: TensorDict) -> torch.Tensor:
+        with torch.no_grad():
+            tensordict = tensordict.select("posterior_state", "belief", "reward")
+
+            tensordict.batch_size = [
+                tensordict.shape[0],
+                tensordict.get("belief").shape[1],
+            ]
+            tensordict.rename_key("posterior_state", "prior_state")
+            tensordict = tensordict.view(-1).detach()
+        with hold_out_net(self.model_based_env), set_exploration_mode("random"):
+            tensordict = self.model_based_env.rollout(
+                max_steps=self.imagination_horizon,
+                policy=self.actor_model,
+                auto_reset=False,
+                tensordict=tensordict,
+            )
+            tensordict = step_tensordict(
+                tensordict,
+                keep_other=True,
```

Comment:
this is dangereous:
say you `next_prior_state` is not there but you have `prior_state`
Then you will think you're working with the `next_prior_state` but you're not.
If would turn this off if possible. Otherwise, can you tell which value needs to be kept?

Reply:
Why would next_prior_state not be there? step tensordict is anyway called between each rollout step so if we don't have this, it would also be confusing.